### PR TITLE
refactor: composition pipeline and qpf body parsing

### PR DESCRIPTION
**Pull request by Equilibris**
Have to discuss if the changes made to `is_trivial` are logically equivalent. Lots of use of bullet should be rewritten, this should probably be done together in the review to see the best way to write them. In some cases, there are bullet free versions below the line. Some changes to pattern matching, it might be worth modifying Comp.parseApp as well to allow for reuse of match results if this results in more readable code.

### CODE REVIEW

**Comment by Equilibris** [comment]
I also added some refactors for Replace in the same pr here, will be looking for more TODOs as wello

File: Qpf/Macro/Data/Replace.lean [hunk 1]
```diff
@@ -32,25 +36,23 @@ private abbrev ReplaceM := StateT Replace m
...
 private def Replace.new : m Replace := 
-  do pure ⟨#[], #[], ⟨#[], #[]⟩⟩
+  do pure ⟨#[], ⟨#[], #[]⟩⟩
 
 private def CtorArgs.reset : ReplaceM m Unit := do
   let r ← StateT.get
   let n := r.vars.size
   let ctor: CtorArgs := ⟨#[], (Array.range n).map fun _ => #[]⟩
-  StateT.set ⟨r.expr, r.vars, ctor⟩
+  StateT.set { r with ctor }
```
Comment (alexkeizer) [inline Qpf/Macro/Data/Replace.lean:45]: This seems like an unrelated change, please put that in it's own PR. 
If you put the trivial stuff separately, I can quickly review and merge that, removing noise from the current diff

**Review by alexkeizer — COMMENTED** [review]
The diff is quite noisy and hard to review.
Generally, when doing big refactors, it's best to not move around *and* change code at the same time.

In this case, could you first submit a PR that takes the current code, and just copies it exactly into separate functions. That will have a large diff, but it should be relatively easy to confirm it's just moving code around and nothing semantically changed.

Then, in a second PR, make the semantic changes. That way, this second diff will be much more focussed, and it will be easier to see the relevant changes

**Comment by Equilibris** [comment]
Closed in favour of #21 #22 and #23

**Pull request closed without merging**